File: executor/agents/claude_code/prompt_enrichment.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""Prompt enrichment helpers for Claude Code requests."""

from typing import Any, Union

from executor.agents.claude_code.multimodal_prompt import append_text_to_vision_prompt

PromptType = Union[str, list[dict[str, Any]]]


def inject_kb_meta_prompt(
    prompt: PromptType,
    kb_meta_prompt: str,
    *,
    executor_mode: str,
    is_user_selected_kb: bool,
    task_type: str | None = None,
) -> PromptType:
    """Prepend KB metadata context for non-code Claude Code requests."""
    if not kb_meta_prompt:
        return prompt

    if (task_type or "").strip().lower() == "code":
        return prompt

    guidance_lines = [
        "<knowledge_base_guidance>",
        "Knowledge base routing:",
    ]
    if is_user_selected_kb:
        guidance_lines.append(
            "- The knowledge base IDs in the context below were selected by the user."
        )
    else:
        guidance_lines.append(
            "- The context below provides knowledge base IDs that may answer the request."
        )
    guidance_lines.extend(
        [
            "- First answer by querying the provided knowledge base ID(s). When a Wegent knowledge tool accepts `knowledge_base_id` or `knowledge_base_ids`, pass the ID(s) from the context.",
            "- If the provided knowledge base ID(s) cannot satisfy the request because results are empty, irrelevant, inaccessible, or incomplete, broaden the query to all knowledge bases when the tool supports it.",
            "- If knowledge base retrieval still cannot answer the request, then use web search or other external tools when available and appropriate.",
            "- When you need to identify which document matters, call `wegent_kb_list_documents` first.",
            "- When you need the content of a specific knowledge base document, call",
            "  `wegent_kb_read_document_content` with `document_id` and optional `offset`/`limit`.",
            "- Do not construct MCP resource URIs manually.",
            "</knowledge_base_guidance>",
        ]
    )
    kb_guidance = "\n".join(guidance_lines) + "\n"

    kb_context = (
        f"{kb_guidance}<knowledge_base_context>\n"
        f"{kb_meta_prompt}\n"
        "</knowledge_base_context>"
    )
    if isinstance(prompt, list):
        return append_text_to_vision_prompt(prompt, kb_context, prepend=True)
    return f"{kb_context}\n\n{prompt}"

File: executor/src/prompt_enrichment.rs
// SPDX-FileCopyrightText: 2026 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

use serde_json::Value;

use crate::attachments::append_text_to_vision_prompt;

pub fn inject_kb_meta_prompt(
    prompt: &Value,
    kb_meta_prompt: &str,
    is_user_selected_kb: bool,
    task_type: Option<&str>,
) -> Value {
    let kb_meta_prompt = kb_meta_prompt.trim();
    if kb_meta_prompt.is_empty() || is_code_task(task_type) {
        return prompt.clone();
    }

    let kb_context = build_kb_context(kb_meta_prompt, is_user_selected_kb);
    match prompt {
        Value::String(text) => Value::String(format!("{kb_context}\n\n{text}")),
        Value::Array(_) => append_text_to_vision_prompt(prompt, &kb_context, true),
        _ => prompt.clone(),
    }
}

fn build_kb_context(kb_meta_prompt: &str, is_user_selected_kb: bool) -> String {
    let mut guidance_lines = vec![
        "<knowledge_base_guidance>".to_owned(),
        "Knowledge base routing:".to_owned(),
    ];
    if is_user_selected_kb {
        guidance_lines.push(
            "- The knowledge base IDs in the context below were selected by the user.".to_owned(),
        );
    } else {
        guidance_lines.push(
            "- The context below provides knowledge base IDs that may answer the request."
                .to_owned(),
        );
    }
    guidance_lines.extend([
        "- First answer by querying the provided knowledge base ID(s). When a Wegent knowledge tool accepts `knowledge_base_id` or `knowledge_base_ids`, pass the ID(s) from the context.".to_owned(),
        "- If the provided knowledge base ID(s) cannot satisfy the request because results are empty, irrelevant, inaccessible, or incomplete, broaden the query to all knowledge bases when the tool supports it.".to_owned(),
        "- If knowledge base retrieval still cannot answer the request, then use web search or other external tools when available and appropriate.".to_owned(),
        "- When you need to identify which document matters, call `wegent_kb_list_documents` first.".to_owned(),
        "- When you need the content of a specific knowledge base document, call".to_owned(),
        "  `wegent_kb_read_document_content` with `document_id` and optional `offset`/`limit`.".to_owned(),
        "- Do not construct MCP resource URIs manually.".to_owned(),
        "</knowledge_base_guidance>".to_owned(),
    ]);

    format!(
        "{}\n<knowledge_base_context>\n{}\n</knowledge_base_context>",
        guidance_lines.join("\n"),
        kb_meta_prompt
    )
}

fn is_code_task(task_type: Option<&str>) -> bool {
    task_type
        .map(str::trim)
        .is_some_and(|value| value.eq_ignore_ascii_case("code"))
}
